File: linkList/92ReverseLinkedListII.py
class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def reverseBetween(self, head, m, n):
        """
        :type head: ListNode
        :type m: int
        :type n: int
        :rtype: ListNode
        """
        # 定义头节点
        pre = ListNode(None)
        pre.next = head
        slow = pre
        # 找到目标节点的前一个节点
        for _ in range(m - 1):
            slow = slow.next
        # 将[m - n] 区间的链表逆置
        sh = slow.next
        s = slow.next
        f = s.next
        for _ in range(n - m):
            t = f.next
            f.next = s
            s = f
            f = t
        # 连接链表
        sh.next = f
        slow.next = s
        return pre.next

File: java/src/LinkList/92.reverse-linked-list-ii.java
/*
 * @lc app=leetcode id=92 lang=java
 *
 * [92] Reverse Linked List II
 *
 * https://leetcode.com/problems/reverse-linked-list-ii/description/
 *
 * algorithms
 * Medium (34.11%)
 * Total Accepted:    184.1K
 * Total Submissions: 536.2K
 * Testcase Example:  '[1,2,3,4,5]\n2\n4'
 *
 * Reverse a linked list from position m to n. Do it in one-pass.
 * 
 * Note: 1 ≤ m ≤ n ≤ length of list.
 * 
 * Example:
 * 
 * 
 * Input: 1->2->3->4->5->NULL, m = 2, n = 4
 * Output: 1->4->3->2->5->NULL
 * 
 * 
 */
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode reverseBetween(ListNode head, int m, int n) {
        if(head == null) {
            return null;
        }
        ListNode dummy = new ListNode(0);
        dummy.next = head;
        ListNode pre = dummy;
        for(int i = 0 ; i<m-1;i++){
            pre = pre.next;
        }
        ListNode start  = pre.next;
        ListNode then = start.next;
        
        for(int j = 0;j<n-m;j++){
            start.next = then.next;
            then.next = pre.next;
            pre.next = then;
            then = start.next;
            
        }
        return dummy.next;
    }
}

